### PR TITLE
Support recursive extras

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

We had a guard in the resolve to avoid "self-dependencies" (as in `gps3==0.33.3`), but this guard was _unintentionally_ filtering out recursive extras.

Closes https://github.com/astral-sh/uv/issues/1342.

## Test Plan

Taken from https://github.com/astral-sh/uv/pull/1352.

### CODE REVIEW

File: crates/uv-resolver/src/pubgrub/dependencies.rs [hunk 1]
```diff
@@ -21,22 +21,16 @@ impl PubGrubDependencies {
         requirements: &[Requirement],
         constraints: &[Requirement],
         overrides: &Overrides,
-        extra: Option<&ExtraName>,
-        source: Option<&PackageName>,
+        source_name: Option<&PackageName>,
+        source_extra: Option<&ExtraName>,
```
Comment (charliermarsh) [inline crates/uv-resolver/src/pubgrub/dependencies.rs:25]: I just renamed and re-ordered these (especially since `extra` got shadowed below).

**Review by zanieb — APPROVED** [review]
Nice, that's the code I was suspicious of

**Comment by zanieb** [comment]
Are these warnings only visible with `-v` or what? Why weren't we seeing them?

**Comment by charliermarsh** [comment]
Yeah

File: crates/uv/tests/pip_install_scenarios.rs [hunk 2]
```diff
@@ -735,6 +735,76 @@ fn multiple_extras_required() {
     assert_installed(&context.venv, "c_502cbb59", "1.0.0", &context.temp_dir);
 }
 
+/// all-extras-required
+///
+/// Multiple optional dependencies are requested for the via an 'all' extra.
```
Comment (jankatins) [inline crates/uv/tests/pip_install_scenarios.rs:740]:
```suggestion
/// Multiple optional dependencies are requested for the package via an 'all' extra.
```

Comment (zanieb) [reply crates/uv/tests/pip_install_scenarios.rs:740]: Thanks! This file is actually generated. If you want to fix it, it's sourced from https://github.com/zanieb/packse/blob/main/scenarios/extras.json

Comment (jankatins) [reply crates/uv/tests/pip_install_scenarios.rs:740]: https://github.com/zanieb/packse/pull/112

**Pull request merged**